pkgs/development/ruby-modules/bundler-env/default.nix

```diff
@@ -67,25 +76,54 @@ let
     if hasBundler then gems.bundler
     else defs.bundler.override (attrs: { inherit ruby; });
 
-  gems = lib.flip lib.mapAttrs configuredGemset (name: attrs:
-    buildRubyGem ((removeAttrs attrs ["source"]) // attrs.source // {
-      inherit ruby;
-      gemName = name;
-      gemPath = map (gemName: gems."${gemName}") (attrs.dependencies or []);
-    }));
+  pathDerivation = {
+    usesGemspec ? false, ...
+  }@attrs:
+    let
+      res = {
+          inherit usesGemspec;
+          type = "derivation";
+          name = attrs.gemName;
+          version = attrs.version;
+          outPath = attrs.path;
+          outputs = [ "out" ];
+          out = res;
+          outputName = "out";
+        };
+    in res;
+
+  buildGem = name: attrs: (
+    let
+      gemAttrs = ((removeAttrs attrs ["source"]) // attrs.source // {
+        inherit ruby;
+        gemName = name;
+        gemPath = map (gemName: gems."${gemName}") (attrs.dependencies or []);
+      });
+    in
+    if gemAttrs.type == "path" then pathDerivation gemAttrs
+    else buildRubyGem gemAttrs);
+
+  gems = lib.flip lib.mapAttrs configuredGemset (name: attrs: buildGem name attrs);
+
+  maybeCopyAll = { usesGemspec ? false, ...}@main:
+  (if usesGemspec then ''
+  cp -a ${gemdir}/* $out/
+  '' else ""
+  );
 
   # We have to normalize the Gemfile.lock, otherwise bundler tries to be
   # helpful by doing so at run time, causing executables to immediately bail
   # out. Yes, I'm serious.
   confFiles = runCommand "gemfile-and-lockfile" {} ''
     mkdir -p $out
-    cp ${gemfile'} $out/Gemfile
-    cp ${lockfile'} $out/Gemfile.lock
+    ${maybeCopyAll mainGem}
+    cp ${gemfile'} $out/Gemfile || ls -l $out/Gemfile
+    cp ${lockfile'} $out/Gemfile.lock || ls -l $out/Gemfile.lock
   '';
 
   envPaths = lib.attrValues gems ++ lib.optional (!hasBundler) bundler;
 
-  binPaths = if mainGem != null then [ mainGem ] else envPaths;
+  binPaths = if mainGem != null then [ mainGem ] ++ envPaths else envPaths;
```

Comment:
The summary is: I think the requirements here are ambiguous and I needed this interpretation.

My primary motivation here has been in cases where I'm developing a gem or ruby app, and so I want to use bundlerEnv in shell.nix. In that case, the Gemfile often includes a bunch of gems that are incltuded for use in development - and if there's an executable in the main gem, I want to be able to run it as well.

This may be in tension with the case where we're installing a ruby app with nix-env, so we don't want to pull in e.g. rspec as a executable. (As an aside, IMO if you're distrubing a ruby executable, then you distribute a gem, so nixpkgs should use buildRubyGem... but that's not even 100% from a ruby-community standpoint.)

I haven't confirmed this, but there's also the significant use case of shipping Rails apps, and they absolutely assume a git repo checked out and bundled in place, so there's definitely a use case for bundlerEnv for applications and therefore avoiding collisions with rspec. That said though, those rails apps will have bin directories with the same set of names...

The upshot is probably that the conditional here shouldn't be "mainGem != null" - but I don't know what the conditional should be.